### PR TITLE
Expose ECDSA sign-to-contract and Anti-Klepto protocol functions

### PR DESCRIPTION
**Pull request by LeoComandini**
Expose functionalities added in https://github.com/ElementsProject/secp256k1-zkp/pull/111

Added a section in "Developer Guides" with an example of the Anti-Klepto protocol.

### CODE REVIEW

File: src/test/test_anti_klepto.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            (pub_key, 32, msg,     32, host_rand, 32, signer_commitment, 33, 0,     sig,     64), # Unsupported flags
+            (pub_key, 32, msg,     32, host_rand, 32, signer_commitment, 33, flags, None,    64), # Missing sig
+            (pub_key, 32, msg,     32, host_rand, 32, signer_commitment, 33, flags, sig,     64), # Incorrect sig length
+            (pub_key, 32, msg,     32, host_rand, 32, signer_commitment, 33, flags, inv_sig, 64), # Invalid sig
+            ]:
+            self.assertEqual(WALLY_EINVAL, wally_anti_klepto_verify(*args))
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (jgriffiths) [inline src/test/test_anti_klepto.py:94]: nice tests!

File: src/test/test_sign.py [hunk 2]
```diff
@@ -244,6 +244,53 @@ def test_recoverable_sig(self):
...
+    def test_s2c(self):
+        priv_key, pub_key, msg, s2c_data, sig_out, s2c_opening_out = self.cbufferize(
+            ['11' * 32, '00' * 32, '22' * 32, '33' * 32, '00' * 64, '00' * 33])
+
+        flags = FLAG_ECDSA
+
+        self.assertEqual(WALLY_OK, wally_s2c_sig_from_bytes(priv_key, 32, msg, 32, s2c_data, 32, flags, sig_out, 64, s2c_opening_out, 33))
+
+        self.assertEqual(WALLY_OK, wally_ec_public_key_from_private_key(
+            priv_key, 32, pub_key, 33))
```
Comment (jgriffiths) [inline src/test/test_sign.py:256]: nit: no need to line break here

File: include/wally_crypto.h [hunk 3]
```diff
@@ -523,6 +523,65 @@ WALLY_CORE_API int wally_ecdh(
...
+ * :param sig_len: The length of ``sig`` in bytes. Must be ``EC_SIGNATURE_LEN``.
+ * :param s2c_data: The data that was committed to.
+ * :param s2c_data_len: The length of ``s2c_data`` in bytes. Must be ``S2C_DATA_LEN``.
+ * :param s2c_opening: The opening information produced during signing.
+ * :param s2c_opening_len: The length of ``s2c_opening`` in bytes. Must be ``S2C_OPENING_LEN``.
+ * :param flags: Must be ``EC_FLAG_ECDSA``.
+ *
+ * .. note:: This function requires external locking if called from multiple threads.
+ */
+WALLY_CORE_API int wally_s2c_verify(
```
[the remark this answers is not in the document]
Comment (LeoComandini) [reply include/wally_crypto.h:576]: This is not verifying the signature, it just verifies the commitment. Do we still want to rename this? Should I mention that in the description?

[the remark this answers is not in the document]
Comment (jgriffiths) [reply include/wally_crypto.h:576]: Ah, I think `wally_s2c_commitment_verify` then, still matching the other verify functions naming convention/

**Comment by LeoComandini** [comment]
Addressed the comments and rearranged the commits, which had some issues.

**Comment by jgriffiths** [comment]
Pretty close to the ideal PR now, thanks!

There's only one issue left that stands out after building locally; the new functions need to be wrapped in the Java `swig.i` file with `returns_xxx` - that part of the Java wrapping hasn't been automated yet.

I will be merging #257 shortly and I've added a test to it that will fail when unwrapped Java calls are present so this gets caught automatically going forward.

**Pull request merged**